### ISSUE TITLE
[Roadmap] v0.3 release checklist

### ISSUE THREAD

**Issue by jermainewang** [dmlc/dgl#450]
Here is the v0.3 release plan. **The tentative release date is 06/07.**

[Feature] Kernel support
-------------------------
Kernels are critical for our system performance. The next release will include all the basic building blocks and APIs for future extensions.
- [x] Basic fused kernel support (#519 )
- [x] New builtins (@jermainewang @ylfdq1118 ) [05/29]
  * `src_op_edge`, `src_op_dst`, `edge_op_dst`, `dst_op_edge`, etc.
- [x] MXNet fused kernel support (#541 @ylfdq1118 ) [05/29]
- [x] Sparse softmax module & GroupEdgeApply [06/01]
- [x] Pip/conda installation (@BarclayII @jermainewang ) [06/06]

[Feature] Giant graph support
---------------------------------------
Release a demo about how to train GNNs on giant graphs that cannot be hosted in a single GPU memory. This includes:
- [x] Sampler and Nodeflow tutorial (#522 @zheng-da )
- [x] Shared-memory graph store and distributed sampler tutorial (@zheng-da @aksnzhy)
- [x] Distributed sampler
  - [x] RPC and network communication (#465 @aksnzhy )
  - [x] Distributed sampler example (#564 #562 @aksnzhy )
- [x] GraphStore APIs (#468 @zheng-da )
- [x] End-to-end single-machine shared-memory based giant graph example. (@zheng-da )

[Enhancement] NN module
---------------------------
- [x] Efficient EdgeSoftmax using new builtins
- [ ] GAT

[Enhancement] graph structure
----------------------------------------
- [x] Refactor immutable graph to have better CSR/COO, context and datatype support (#543 @jermainewang )
- [x] Batching immutable graphs (#539 @VoVAllen )
- [x] Random walk APIs (#453 @BarclayII )
- [x] Support edge/node removal (Issue #467 @yzh119 PR #599 )
- [x] Support convenient `DGLGraph.to(device)` (Issue #503 @HQ01  PR #600 )
- [x] `dgl.to_simple_graph` API to remove multi-edges (#587 @jermainewang )
- [x] `dgl.to_bidirected` API to transform the graph to bidrectional. (#598 @mufeili )

Model Examples
----------------------
- [x] APPNP (#480 @aymenwah )
- [x] GIN (#471 @kitaev-chen )
- [x] PinSage (#453 @BarclayII )

Bug fix
------------------
- [x] NodeFlow bugs (#531 @zheng-da )
- [x] NumPy bug (Issue #538 PR #580  @jermainewang )

## Features postponed to v0.4

**[Feature] DGL graph data format**
As we want to include more popular graph dataset in DGL, it is time to decouple dataset with DGL repo.
* Design a unified graph data format for DGL data loader. The data format should be (1) efficient for read/write; (2) easy for partitioning
* Fast data loader in C++.
* Convert our curated dataset to such format.
* More dataset?

**[Feature] Bipartite (k-partite) graph API**
Bipartite graph is popular in recommendation setting. We heard many requests for this.

**[Enhancement] Improve IR & scheduling system**
* Improve pull by using slicing on CSR graph.

**Comment by yzh119** [comment]
Some additional items:
- Kernel Support
  - sparse_softmax
- Convenient RandomWalk api
  - to support algorithms like personalized page rank.
- Model Examples
  - relational network/recurrent relational network/relational deep reinforcement learning(however this one is much more difficult). 
  - these models are good showcases as using graph to modeling text/image.

**Comment by yzh119** [comment]
For API, I suggest provide more flexible message function to handle different type of edges.
For example, In tree-lstm we are supposed to send different messages for left branch and right branch.

**Comment by aksnzhy** [comment]
We plan to implement a light-weight kvstore instead of using MXNet KVStore for DGL. Do we consider add this item to 0.3 version?

**Comment by zheng-da** [comment]
- kernel support:
  - scatter mean
  - generalized spmv: min/max/mean for reduce.

- dataset:
  - [reddit](https://files.pushshift.io/reddit/comments/): we construct multiple versions at different scales and different ways of construction (mainly how to build edges between nodes).
  - [yelp](https://www.yelp.com/dataset/documentation/main)

**Comment by zheng-da** [comment]
> We plan to implement a light-weight kvstore instead of using MXNet KVStore for DGL. Do we consider add this item to 0.3 version?

0.3 version will be a fast release. How about we plan it for 0.4 release?

**Comment by aksnzhy** [comment]
> > We plan to implement a light-weight kvstore instead of using MXNet KVStore for DGL. Do we consider add this item to 0.3 version?
> 
> 0.3 version will be a fast release. How about we plan it for 0.4 release?

No problem.

**Comment by qzshadow** [comment]
Looking forward to supports for giant graphs!

**Comment by zheng-da** [comment]
Scheduling improvement:
* we should improve pull. (It's commonly used by control-variate sampling and we run pull on the giant graph.)

**Comment by zheng-da** [comment]
If time permits, we might add GPU samplers to accelerate mini-batch training in a single GPU.

**Comment by alexvpickering** [comment]
A couple models I'd like to see:

- [Graph Isomorphism Network](https://github.com/weihua916/powerful-gnns)
- [LanczosNet](https://github.com/lrjconan/LanczosNetwork)

**Comment by jermainewang** [comment]
Hi @alexvpickering , will look into it. One question, do you expect to see them as layers in NN packages or complete examples with data loading/training/testing ?

**Comment by alexvpickering** [comment]
I was thinking examples like found in in e.g.  [dgl/examples/pytorch/sgc](https://github.com/dmlc/dgl/tree/master/examples/pytorch/sgc).

**Comment by HuangZhanPeng** [comment]
will some model in [Euler](https://github.com/alibaba/euler) like LsHNE, LasGNN, ScalableGCN implement in the feature?

**Comment by yzh119** [comment]
@HuangZhanPeng are these algorithms any special?

**Comment by zheng-da** [comment]
@HuangZhanPeng We will work on heterogeneous graphs in the 0.4 release. We'll evaluate these models.

**Comment by klicperajo** [comment]
It would be great if you could also implement [APPNP](https://github.com/klicperajo/ppnp)! It's quite simple and performed best in PyTorch Geometric's benchmark, so people could clearly benefit here as well. :)

**Comment by mufeili** [comment]
Thank you for the suggestion @klicperajo . I think we've already had one implementation for this work:). See [dgl APPNP example](https://github.com/dmlc/dgl/tree/master/examples/pytorch/appnp).

**Comment by klicperajo** [comment]
Oh, nice! I didn't notice since it's not in the summary table. :)

**Comment by jermainewang** [comment]
Just updated the roadmap with a checklist. Our tentative date for this release is **05/07.**

For all committers @zheng-da @szha @BarclayII @VoVAllen @ylfdq1118 @yzh119 @GaiYu0 @mufeili @aksnzhy @zzhang-cn @ZiyueHuang , please vote with :+1:  if you agree with this plan.

**Comment by szha** [comment]
did you mean 6/7?

**Comment by VoVAllen** [comment]
@yzh119 will be in charge of edge/node removal

**Comment by mufeili** [comment]
@jermainewang What do you need for issues under NN module?

**Comment by yzh119** [comment]
One question: do we need to rebuild node/edge index when calling node/edge removal?
For example:
```
import dgl
import torch as th
g = dgl.DGLGraph()
g.add_nodes(5)
g.ndatas['x'] = th.rand(5, 3)
g.del_nodes([2, 3])
print(g.nodes())
```
What should be the output? `tensor([0, 1, 4])`?

**Comment by yzh119** [comment]
I must say, if the node index and edge index must be rebuilt after node/edge removal, the behavior of these operations would be VERY CONFUSING. I can't see any scenario that these operations would make any sense.

**Comment by VoVAllen** [comment]
I think we need to return the removed edges, and let user handle the mapping between original index and modified index. (or provide utils function)

**Comment by yzh119** [comment]
@VoVAllen , I don't see any benefits of doing this compared to creating subgraph of the current graph.

**Comment by jermainewang** [comment]
> @jermainewang What do you need for issues under NN module?

(1) EdgeSoftmax needs to be re-implemented using the new builtins; (2) GAT as an NN module. Both depend on the kernel branch to be fully merged.

@mufeili do you have time take over these two items?
* `dgl.to_simple_graph` API to remove multi-edges
* `dgl.to_bidirected` API to transform the graph to bidrectional. (Issue #419)

**Comment by mufeili** [comment]
@jermainewang I may need to wait till weekend for implementing them. If that's good for you, I'll take over them.

**Comment by jermainewang** [comment]
I've taken `to_simple_graph`. The rest is yours.

**Comment by jermainewang** [comment]
v0.3 has been released. Thanks everyone!
